HACKING.md

```diff
@@ -0,0 +1,103 @@
+## Installing Eio from Git
+
+If you want to run the latest development version from Git, run these commands:
+
+```
+git clone https://github.com/ocaml-multicore/eio.git
+cd eio
+opam pin -yn .
+opam install eio_main
+```
+
+## Layout of the code
+
+`lib_eio/core` contains the core logic about fibers, promises, switches, etc.
+`lib_eio` extends this with e.g. streams, buffered readers, buffered writers,
+and a load of types for OS resources (files, networks, etc).
+
+There is one directory for each backend (e.g. `eio_linux`).
+Each backend provides a scheduler that integrates with a particular platform,
+and implements some or all of the cross-platform resource APIs.
+For example, `eio_linux` implements the network interface using `io_uring` to send data.
+
+`lib_main` just selects an appropriate backend for the current system.
+
+## Writing a backend
+
+It's best to start by reading `lib_eio/mock/backend.ml`, which implements a mock backend with no actual IO.
+You can then read one of the real backends to see how to integrate this with the OS.
+
+Most backends are built in two layers:
+
+- A "low-level" module directly wraps the platform's own API, just adding support for suspending fibers for concurrency
+  and basic safety features (such wrapping `Unix.file_descr` to prevent use-after-close races).
+
+- An implementation of the cross-platform API defined in the `eio` package that uses the low-level API internally.
+  This should ensure that errors are reported using the `Eio.Io` exception.
+
+## Tests
+
+Eio has tests in many places...
+
+### Cross-platform unit tests
+
+These are in the top-level `tests` directory.
+They are run against whichever backend `Eio_main.run` selects, and therefore must get the same result for all backends.
+
+### Concurrency primitives
+
+`lib_eio/tests` tests some internal data structures, such as the lock-free cells abstraction.
+The `.md` files in that directory provide a simple walk-through to demonstrate the basic operation,
+while `lib_eio/tests/dscheck` uses [dscheck][] to perform exhaustive testing of all atomic interleavings
```

Comment:
Broken link here?

Reply:
It works for me.

Reply:
I'm using the preview at https://github.com/ocaml-multicore/eio/blob/e332fdbfec500f7ac6f4380f32278f734053990c/HACKING.md#concurrency-primitives - does that work for you?

Reply:
Yes, it does.   I didn't know about reference-style links &mdash; and now that I do, I notice they are actually documented on a page I'm pretty sure I've read before on markdown!  What is this magic, a heisenfeature?  😅

Carry on!  LGTM